File: src/main/java/hzt/aoc/day15/Part2RambunctiousRecitation.java
package hzt.aoc.day15;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Part2RambunctiousRecitation extends Day15Challenge {

    public Part2RambunctiousRecitation() {
        super("part 2",
                "Given your starting numbers, what will be the 30000000th number spoken?");
    }

    private static final int THRESHOLD = 30_000_000;
    private static final int LOG_STEP = 3_000_000;

    @Override
    protected int getNthNumberSpoken(List<Integer> numbers) {
        long start = System.nanoTime();
        int index = 0;
        int last = -1;
        Map<Integer, Integer> seenLastToIndex = new HashMap<>();
        for (int i = 0; i < numbers.size(); i++) {
            index++;
            last = numbers.get(i);
            if (i != numbers.size() - 1) seenLastToIndex.put(last, index);
        }
        while (index < THRESHOLD) {
            int seenLastTemp = seenLastToIndex.getOrDefault(last, -1);
            seenLastToIndex.put(last, index);
            last = seenLastTemp == -1 ? 0 : (index - seenLastTemp);
            index++;
            start = logTime(index, LOG_STEP, 0, last, start);
        }
        return last;
    }

    @Override
    protected String getMessage(String global) {
        return String.format("The %dth number spoken is: %s", THRESHOLD, global);
    }
}

File: src/main/java/hzt/aoc/day15/Part2RambunctiousRecitation.kt
package hzt.aoc.day15

import java.util.HashMap
import java.lang.System

// Credits to TurkeyDev
class Part2RambunctiousRecitation : Day15Challenge(
    "part 2",
    "Given your starting numbers, what will be the 30000000th number spoken"
) {
    override fun getNthNumberSpoken(numbers: MutableList<Int>): Int {
        var start = System.nanoTime()
        var index = 0
        var last = -1
        val seenLastToIndex: MutableMap<Int, Int> = HashMap()
        for (i in numbers.indices) {
            index++
            last = numbers[i]
            if (i != numbers.size - 1) seenLastToIndex[last] = index
        }
        while (index < THRESHOLD) {
            val seenLastTemp = seenLastToIndex.getOrDefault(last, -1)
            seenLastToIndex[last] = index
            last = if (seenLastTemp == -1) 0 else index - seenLastTemp
            index++
            start = logTime(index, LOG_STEP, 0, last, start)
        }
        return last
    }

    override fun getMessage(global: String): String {
        return String.format("The %dth number spoken is: %s", THRESHOLD, global)
    }

    companion object {
        private const val THRESHOLD = 30000000
        private const val LOG_STEP = 3000000
    }
}
